secretflow/ml/nn/sl/backend/tensorflow/strategy/pipeline.py

```diff
@@ -0,0 +1,217 @@
+            eval_data = next(self.eval_set)
+            if self.eval_has_y:
+                if self.eval_has_s_w:
+                    data_x = eval_data[:-2]
+                    eval_y = eval_data[-2]
+                    self.eval_sample_weight = eval_data[-1]
+                else:
+                    data_x = eval_data[:-1]
+                    eval_y = eval_data[-1]
+                # Label differential privacy
+                if self.label_dp is not None:
+                    dp_eval_y = self.label_dp(eval_y.numpy())
+                    self.eval_y = tf.convert_to_tensor(dp_eval_y)
+                else:
+                    self.eval_y = eval_y
+            else:
+                data_x = eval_data
+        else:
+            raise Exception("invalid stage")
+
+        # Strip tuple of length one, e.g: (x,) -> x
+        data_x = data_x[0] if isinstance(data_x, Tuple) and len(data_x) == 1 else data_x
+
+        tape = tf.GradientTape(persistent=True)
+        if stage == "train":
+            self.base_tape.append(tape)
+            self.trainable_vars.append(self.model_base.trainable_variables)
+        with tape:
+            h = self._base_forward_internal(
+                data_x,
+            )
+        if stage == "train":
+            self.h.append(h)
+
+        self.data_x = data_x
+
+        forward_data = ForwardData()
+        if len(self.model_base.losses) > 0:
+            forward_data.losses = tf.add_n(self.model_base.losses)
+        # TODO: only vaild on no server mode, refactor when use agglayer or server mode.
+        # no need to compress data on model_fuse side
+        if compress and not self.model_fuse:
+            if self.compressor:
+                # forward_data.hidden = self.compressor.compress(self.h.numpy())
+                forward_data.hidden = self.compressor.compress(h.numpy())
+            else:
+                raise Exception(
+                    'can not find compressor when compress data in base_forward'
+                )
+        else:
+            forward_data.hidden = h
+        return forward_data
+
+    def base_backward(self, gradient, compress: bool = False):
+        """backward on fusenet
+
+        Args:
+            gradient: gradient of fusenet hidden layer
+            compress: Whether to decompress gradient.
+        """
+        return_hiddens = []
+
+        # TODO: only vaild on no server mode, refactor when use agglayer or server mode.
+        # no need to decompress data on model_fuse side
+        if compress and not self.model_fuse:
+            if self.compressor:
+                gradient = self.compressor.decompress(gradient)
+            else:
+                raise Exception(
+                    'can not find compressor when decompress data in base_backward'
+                )
+
+        assert (
+            len(self.base_tape) == self.pipeline_size
+        ), f'mismatch pipeline, pipeline size: {self.pipeline_size-1}, base_tape: {len(self.base_tape)}'
```

Comment:
`{self.pipeline_size-1}` -> `{self.pipeline_size}`

Reply:
Removed `base_tape` check and add `hiddens_buf` check in sl_model.py